docs/DeveloperGuide.md

```diff
@@ -346,14 +415,229 @@ Priorities: High (must have) - `* * *`, Medium (nice to have) - `* *`, Low (unli
+
+      Use case resumes from step 4.
+
+* 4a. The given format is invalid.
+
+    * 4a1. ContactCS requests the user to provide the correct format and shows the valid command format
+    * 4a2. User enters new input
+    * Steps 4a1 - 4a2 are repeated until the format for the entered input is correct
+
+      Use case resumes from step 4.
+
+**Use case: Access frequently used contacts**
+
+**MSS**
+
+1.  User requests to access frequently used contacts
+2.  ContactCS shows a list of persons
+
+    Use case ends.
+
+**Extensions**
+
+* 2a. The list is empty.
+  * 2a1. ContactCS displays a message to the user saying that the list is empty
+
+    Use case ends.
+
+* 3a. The given format is invalid.
+
+    * 3a1. ContactCS requests the user to provide the correct format and shows the valid command format
+    * 3a2. User enters new input
+    * Steps 3a1 - 3a2 are repeated until the format for the entered input is correct
+
+      Use case resumes from step 2.
+
+**Use case: View usage instructions**
+
+**MSS**
+
+1.  User requests to view usage instructions for the app
+2.  ContactCS displays the usage instructions and commands available for the user
+
+    Use case ends.
+
+**Extensions**
+
+* 2a. The given format is invalid.
+
+    * 2a1. ContactCS requests the user to provide the correct format and shows the valid command format
+    * 2a2. User enters new input
+    * Steps 2a1 - 2a2 are repeated until the format for the entered input is correct
+
+      Use case resumes from step 2.
+
+**Use case: View app with sample data**
+
+**MSS**
+
+1.  User requests to view app with sample data
+2.  ContactCS populates the interface with sample data for demonstration purposes
+3.  User interacts with the app to see how it functions with the sample data
+
+    Use case ends.
+
+**Extensions**
+
+* 2a. The given format is invalid.
+
+    * 2a1. ContactCS requests the user to provide the correct format and shows the valid command format
+    * 2a2. User enters new input
+    * Steps 2a1 - 2a2 are repeated until the format for the entered input is correct
+
+      Use case resumes from step 2.
+
+**Use case: Undo actions**
+
+**MSS**
+
+1.  User performs an action in ContactCS
+2.  User requests to undo the last action
+3.  ContactCS reverts to the state before the last action
+
+    Use case ends.
+
+**Extensions**
+
+* 2a. No action has been performed yet.
+    * 2a1. ContactCS shows an error message indicating that there is no action to undo
+
+      Use case ends.
+
+* 3a. The given format is invalid.
+
+    * 3a1. ContactCS requests the user to provide the correct format and shows the valid command format
+    * 3a2. User enters new input
+    * Steps 3a1 - 3a2 are repeated until the format of the entered input is correct
+
+      Use case resumes from step 3.
+
+**Use case: Redo actions**
+
+**MSS**
+
+1.  User performs an action in ContactCS
+2.  User requests to undo the last action
+3.  ContactCS reverts to the state before the last action
```

Comment:
May be can include the previous use case of undoing task instead of reiterate here again? Like below:
![image](https://github.com/user-attachments/assets/e5be5231-648c-4629-bf56-99df3bd4e8e5)